In [1]:
import pandas as pd
%matplotlib inline

In [12]:
!pwd
!ls ../viz
!ls ../viz/LondonDf.csv

/home/vagrant/git/DS420-Cobras/notebooks
Beijing_AirQuality_Stations_en.xlsx  London_AirQuality_Stations.csv
Beijing_aq.csv			     London_aq.csv
BeijingDf.csv			     LondonDf.csv
Beijing_Env.csv			     London_Env.csv
Beijing_grid_weather_station.csv     London_grid_weather_station.csv
Beijing_Nearest_Neighbor.csv	     London_Nearest_Neighbor.csv
Beijing_Neighbors.xlsx		     London_Neighbors.xlsx
Beijing_points.csv		     London_points.csv
Beijing_weather.csv		     London_weather.csv
GeoSpatialViz.pbix		     stash
../viz/LondonDf.csv


In [20]:
LondonDF = pd.read_csv("../viz/LondonDf.csv")
BeijingDF = pd.read_csv("../viz/BeijingDf.csv")

In [21]:
LondonDF.head(3)

CO_Concentration  NO2_Concentration  O3_Concentration  PM10_Concentration  \
0               NaN               70.4               NaN                14.3   
1               NaN               44.8               NaN                 8.8   
2               NaN               16.5               NaN                 8.9   

   PM25_Concentration  SO2_Concentration       id station_id test_id  \
0                 9.9                NaN  2941506        CD1    None   
1                 8.5                NaN  2941507        BL0    None   
2                 4.8                NaN  2941508        GR4    None   

                  time    ...      precipProbability  precipType pressure  \
0  2018-03-31 07:00:00    ...                    NaN        rain   998.08   
1  2018-03-31 07:00:00    ...                    NaN        rain   998.07   
2  2018-03-31 07:00:00    ...                    NaN        rain   998.00   

         summary  temperature uvIndex  visibility  windBearing  windGust  \
0  Mostly Cloudy         7.23     NaN        7.09          302       NaN   
1  Mostly Cloudy         7.20     NaN        6.88          297       NaN   
2  Mostly Cloudy         7.12     NaN        6.23          284       NaN   

   windSpeed  
0       4.57  
1       4.71  
2       5.55  

[3 rows x 30 columns]

In [22]:
BeijingDF.head(3)

CO_Concentration  NO2_Concentration  O3_Concentration  PM10_Concentration  \
0               0.8               53.0             127.0               172.0   
1               0.9               54.0             121.0               123.0   
2               0.8               66.0             123.0               139.0   

   PM25_Concentration  SO2_Concentration       id   station_id test_id  \
0               105.0               14.0  2941450    dongsi_aq    None   
1                95.0               15.0  2941451   tiantan_aq    None   
2                95.0               13.0  2941452  guanyuan_aq    None   

                  time    ...      precipProbability  precipType pressure  \
0  2018-03-31 07:00:00    ...                    NaN        rain      NaN   
1  2018-03-31 07:00:00    ...                    NaN        rain      NaN   
2  2018-03-31 07:00:00    ...                    NaN        rain      NaN   

   summary  temperature uvIndex  visibility  windBearing  windGust  windSpeed  
0    Clear        12.87     NaN        6.69         60.0       NaN       0.89  
1    Clear        12.88     NaN        6.66         60.0       NaN       0.89  
2    Clear        12.82     NaN        6.92         60.0       NaN       0.89  

[3 rows x 30 columns]

In [18]:
def check_nulls(df):
    for i, column_label in enumerate(df.columns):
        print('{:2d}: {:20} {:6d} unique values, {:6d} null values, {:2d} data type(s)'.format(
            i,
            column_label + ':', 
            df[column_label].nunique(),
            df[column_label].isnull().sum(),
            len(df[column_label].apply(lambda x: type(x)).value_counts())))

In [19]:
check_nulls(LondonDF)

 0: CO_Concentration:         0 unique values,  23022 null values,  1 data type(s)
 1: NO2_Concentration:     1306 unique values,   7318 null values,  1 data type(s)
 2: O3_Concentration:         0 unique values,  23022 null values,  1 data type(s)
 3: PM10_Concentration:     773 unique values,   6054 null values,  1 data type(s)
 4: PM25_Concentration:     697 unique values,   4406 null values,  1 data type(s)
 5: SO2_Concentration:        0 unique values,  23022 null values,  1 data type(s)
 6: id:                   23022 unique values,      0 null values,  1 data type(s)
 7: station_id:              19 unique values,      0 null values,  1 data type(s)
 8: test_id:                625 unique values,      0 null values,  1 data type(s)
 9: time:                  1220 unique values,      0 null values,  1 data type(s)
10: apparentTemperature:   2625 unique values,      0 null values,  1 data type(s)
11: cloudCover:             101 unique values,   1657 null values,  1 data type(s)
12: 

In [23]:
check_nulls(BeijingDF)

 0: CO_Concentration:        43 unique values,   1713 null values,  1 data type(s)
 1: NO2_Concentration:      194 unique values,   1734 null values,  1 data type(s)
 2: O3_Concentration:       319 unique values,   1934 null values,  1 data type(s)
 3: PM10_Concentration:     646 unique values,   8063 null values,  1 data type(s)
 4: PM25_Concentration:     329 unique values,   2183 null values,  1 data type(s)
 5: SO2_Concentration:       79 unique values,   1758 null values,  1 data type(s)
 6: id:                   41090 unique values,      0 null values,  1 data type(s)
 7: station_id:              35 unique values,      0 null values,  1 data type(s)
 8: test_id:               1681 unique values,      0 null values,  1 data type(s)
 9: time:                  1174 unique values,      0 null values,  1 data type(s)
10: apparentTemperature:   3634 unique values,      0 null values,  1 data type(s)
11: cloudCover:             101 unique values,   8153 null values,  1 data type(s)
12: 

## Key Points
**London has 23022 unique values (id has 23022 values)**
**London nulls**
 0: CO_Concentration:         0 unique values,  23022 null values,  1 data type(s)
 1: NO2_Concentration:     1306 unique values,   7318 null values,  1 data type(s)
 2: O3_Concentration:         0 unique values,  23022 null values,  1 data type(s)
 3: PM10_Concentration:     773 unique values,   6054 null values,  1 data type(s)
 4: PM25_Concentration:     697 unique values,   4406 null values,  1 data type(s)
 5: SO2_Concentration:        0 unique values,  23022 null values,  1 data type(s)
 6: id:                   23022 unique values,      0 null values,  1 data type(s)
 7: station_id:              19 unique values,      0 null values,  1 data type(s)
 8: test_id:                625 unique values,      0 null values,  1 data type(s)
 9: time:                  1220 unique values,      0 null values,  1 data type(s)
10: apparentTemperature:   2625 unique values,      0 null values,  1 data type(s)
11: cloudCover:             101 unique values,   1657 null values,  1 data type(s)
12: datetime:              1220 unique values,      0 null values,  1 data type(s)
13: dewPoint:              1392 unique values,      0 null values,  1 data type(s)
14: humidity:                75 unique values,      0 null values,  1 data type(s)
15: icon:                     8 unique values,      0 null values,  1 data type(s)
16: lat:                     20 unique values,      0 null values,  1 data type(s)
17: long:                    19 unique values,      0 null values,  1 data type(s)
18: ozone:                 5876 unique values,  10127 null values,  1 data type(s)
19: precipIntensity:        536 unique values,  10127 null values,  1 data type(s)
20: precipProbability:       88 unique values,  10127 null values,  1 data type(s)
21: precipType:               1 unique values,   8711 null values,  2 data type(s)
22: pressure:              3381 unique values,      0 null values,  1 data type(s)
23: summary:                 12 unique values,      0 null values,  1 data type(s)
24: temperature:           2306 unique values,      0 null values,  1 data type(s)
25: uvIndex:                  7 unique values,  10203 null values,  1 data type(s)
26: visibility:             912 unique values,     13 null values,  1 data type(s)
27: windBearing:            360 unique values,      0 null values,  1 data type(s)
28: windGust:              2849 unique values,  10203 null values,  1 data type(s)
29: windSpeed:             1897 unique values,      0 null values,  1 data type(s)
 